Added switch for generation of interfaces for object and input types

marceloverdijk opened the pull request:
This PR adds a switch for the generation of interfaces for object and input types.
It currently only covers the Java code generator.

I will create it as a draft PR so it can be used as starting point to discuss, as I'm not sure this covers all cases.

Feedback welcome.

marceloverdijk commented:
Thinking about it, having an interface for input object types does actually not make sense. I think they should only be generated for object types with the switch on.

marceloverdijk commented:
@srinivasankavitha I'm looking into this but I could maybe use some help...

Assumptions (but please let me know if you have other thoughts):

- Generating of interfaces is only needed for object types, _not_ for input types (this is an important assumption as the object types and input types are generated the same way using the same code)
- Nested object types should use the interfaces (I will add a example below)
- Should example datafetchers could also use the interfaces? I'm not sure to be honest.
- What about Kotlin? Data types are generated as `data class` and I think that will cause an issue to implement an interface, or am I wrong?

Example:

```
type Query {
    person(id: ID): Person
    people(filter: PersonFilter): [Person]
}

type Person {
    id: ID
    name: String
    email: String
    address: Address
}

type Address {
    street: String
}

input PersonFilter {
    name: String
    email: String
}
```

with `generate-interfaces`=`true` this would generate:

```
public interface IPerson {
    String getId();
    String getName();
    String getEmail();
    IAddress getAddress(); // return interface
}

public interface IAddress {
    String getStreet();
}

public class Person implements IPerson {
    private String id;
    private String name;
    private String email;
    private IAddress address; // use interface
    // getters + setters
}

public class Address implements IAddress {
    private String street;
    // getters + setters
}

public class PersonFilter {
    private String name;
    private String email;
    // getters + setters
}
```

As the same [generate](https://github.com/Netflix/dgs-codegen/blob/master/graphql-dgs-codegen-core/src/main/kotlin/com/netflix/graphql/dgs/codegen/generators/java/DataTypeGenerator.kt#L85) method is used for generating the object and input types we could add a `generateInterface` argument (only `true` for generating object types) to distinguish the generation process.

What do you think?

srinivasankavitha commented:
The primary use case for generating interfaces is to allow other non-generated classes to be used in the data-fetchers, correct? 
For your questions:
1. Generating of interfaces is only needed for object types, not for input types (this is an important assumption as the object types and input types are generated the same way using the same code)
[Kavitha] I agree that it doesn't make sense to add this for input types
2. Nested object types should use the interfaces (I will add a example below)
[Kavitha] yes, I see that this is essential for the feature to work. 
3. Should example datafetchers could also use the interfaces? I'm not sure to be honest.
[Kavitha] Honestly, we haven't seen much usage of the example datafetchers. In most cases, users will need to custom implement them anyway, so I think we can avoid updating the data fetcher generation logic.
4. What about Kotlin? Data types are generated as data class and I think that will cause an issue to implement an interface, or am I wrong?
[Kavitha] Kotlin supports interfaces as well, and yes, this will require changes to the way the kotlin classes are generated. I think that may impact other kotlin users, but not sure unless we try it out.

@paulbakker - thoughts?

marceloverdijk commented:
@srinivasankavitha @paulbakker I've pushed a working solution for the Java code generation.
I'm not 100% happy with changes so maybe good to review this before we move to the Kotlin code generation.
Any comments welcome.

paulbakker commented:
1. Generating of interfaces is only needed for object types, not for input types (this is an important assumption as the object types and input types are generated the same way using the same code)
[Kavitha] I agree that it doesn't make sense to add this for input types
[Paul] +1
2. Nested object types should use the interfaces (I will add a example below)
[Kavitha] yes, I see that this is essential for the feature to work. 
[Paul] +1
3. Should example datafetchers could also use the interfaces? I'm not sure to be honest.
[Kavitha] Honestly, we haven't seen much usage of the example datafetchers. In most cases, users will need to custom implement them anyway, so I think we can avoid updating the data fetcher generation logic.
[Paul] Yeah, I don't think anyone is using that. I'm considering removing generating datafetchers completely, it's just not that useful IMO.
4. What about Kotlin? Data types are generated as data class and I think that will cause an issue to implement an interface, or am I wrong?
[Kavitha] Kotlin supports interfaces as well, and yes, this will require changes to the way the kotlin classes are generated. I think that may impact other kotlin users, but not sure unless we try it out.
[Paul] Although Kotlin obviously does have interfaces, you _can't_ use them with data classes. Since we generate data classes (and should) this makes less sense for Kotlin. I would be ok making this Java only.

As for the `JsonTypeInfo`. That's required so you can descerialize correctly when working with interfaces, so whenever we find an interface type we add these annotations. 

@marceloverdijk once you think you're ready, we should set up some test projects to test it out for real.

marceloverdijk commented:
Hi @paulbakker thx for the feedback!

I think we all agree about point 1. - 3.. 
The other day I was even thinking that there should be a switch to at least switch of the generation of the example datafetchers, but as you are considering removing them all together it's not worth it. For Kotlin they are not generated anyway.

For point 4. as @srinivasankavitha pointed out Kotlin data classes do support interfaces and I think something like this could be generated:

```
interface IPerson {
    val id: String
    val name: String
    val email: String
    val address: IAddress
}

interface IAddress {
    val street: String
}

data class Person(
        override val id: String,
        override val name: String,
        override val email: String,
        override val address: IAddress
) : IPerson

data class Address(
        override val street: String
) : IAddress
```

and an _internal_ implementation with `someInternalProp` could then look like:

```
data class PersonInternal(
        override val id: String,
        override val name: String,
        override val email: String,
        override val address: IAddress,
        val someInternalProp: String
) : IPerson
```

Personally I'm considering moving a (hobby) project to Kotlin so having interfaces for Kotlin as well would be a plus for me.

For the Java code generator I think the code changes a feature complete (maybe it could be implemented a bit differently without passing booleans) and ready for a good test in (test) projects as you mentioned.

Note that I added a quite extensive [unit test](https://github.com/Netflix/dgs-codegen/blob/57bdf48bdb31ae85943ffbb35dba45022780fd2f/graphql-dgs-codegen-core/src/test/kotlin/com/netflix/graphql/dgs/codegen/CodeGenTest.kt#L1226-L1347) as part of the PR. It test object types (including nesting), input types and extensions as well. And all existing unit test are still green. But yes I'm not sure if it covers everything, so a full review and using it some existing or test project is very important.
I think we could start testing it in Java projects and if it works well then look at Kotlin next.

I don't have a big or complex schema to test it with except the existing unit tests and the one I added.
What would you suggest to move forward?

srinivasankavitha commented:
So for kotlin, as @paulbakker suggested, you can use companion classes to add extra fields to a generated class very easily: https://kotlinlang.org/docs/object-declarations.html#companion-objects

We prefer using data classes for Kotlin (as do a lot of our existing users), so adding the support for interfaces would prevent us from doing so. 

The approach of using a flag looks fine to me. For testing, the unit test you have does look good.  Additionally, we do have to take the extra step of applying the plugin in an actual project to verify it works as expected.

marceloverdijk commented:
@srinivasankavitha my Kotlin example in my previous reply contained data classes (similar as the ones generated now) but also implementing an interface. Is there an issue with that? I have only basic Kotlin experience so maybe I'm missing something.

For testing this in a real project is that something you or @paulbakker can arrange or do you expect anything from me?

srinivasankavitha commented:
You're right. It does look possible to use data class with an interface. I was not aware of that.

For testing, we will test this on our internal projects as well. If you could verify the same with your projects, that would be helpful. The reason is the unit test coverage doesn't quite guarantee everything works end to end. You need to use it in a project that consumes the plugin, and uses the generated classes (especially for Kotlin).

marceloverdijk commented:
Today I've run a local snapshot build of the codegen plugin against a schema I have.
After fixing an issue in the PR regarding enums it generates what I expect. But my schema is not that advanced, e.g. no unions, extensions etc.

marceloverdijk commented:
Some other thoughts I was thinking about today.
Let's say the following interfaces are generated:

```
public interface IBook {
    String getTitle();
}

public interface IBookPage {
    List<IBook> getItems();
    IPageInfo getPageInfo();
}
```

I'm wondering if it would be more practical to generate the latter one as:

```
public interface IBookPage {
    List<? extends IBook> getItems();
    IPageInfo getPageInfo();
}
```

Note the the `List<? extends IBook>`

This would allow internal DTO's to be:

```
public class IBookPageInternal implements IBookPage {
    private List<BookInternal> books;
    private IPageInfo pageInfo;

    @Override
    public List<BookInternal> getBooks() {
        return books;
    }

    @Override
    public IPageInfo getPageInfo() {
        return books;
    }    
}
```

Without `List<? extends IBook>`, so with `List<IBook>` instead, implementations *have* to return `List<IBook>` as well.
With `List<? extends IBook>` they can decide what to return and so this could be the internal `BookInternal` making it fully type safe.

srinivasankavitha commented on graphql-dgs-codegen-core/src/main/kotlin/com/netflix/graphql/dgs/codegen/generators/java/DataTypeGenerator.kt:
```diff
@@ -54,8 +54,8 @@ class DataTypeGenerator(private val config: CodeGenConfig, private val document:
                     .filterSkipped()
-                    .map { Field(it.name, typeUtils.findReturnType(it.type, true), isInterfaceType = true) }
-                    .plus(extensions.flatMap { it.fieldDefinitions }.filterSkipped().map { Field(it.name, typeUtils.findReturnType(it.type, true)) })
+                    .map { Field(it.name, typeUtils.findReturnType(it.type, true), overrideGetter = true) }
```
overrideGetter seems to indicate that this intends to add override to the getter only. Is that correct?

marceloverdijk replied:
Yes that's correct. We only need to apply `@Override` to the getter method (as the interface will only have the getter; and the generated implementation class has the property, getter and setter). As alternative we could name that argument `implementsInterface` so it is more general, but currently it's only needed for the getter.

srinivasankavitha commented:
For your example with respect to interface generation - what you describe sounds more user-friendly and convenient. I'd be in favor of adding that if it doesn't complicate the implementation too much. 

As for the testing on our side, will probably get to it in a week or so. Thanks for the changes.

marceloverdijk commented:
> For your example with respect to interface generation - what you describe sounds more user-friendly and convenient. I'd be in favor of adding that if it doesn't complicate the implementation too much.

Then we should add it as imo it has a real user benefit. I already briefly looked into it yesterday and yes likely it will complicate things a bit. I will have a try this week.

> As for the testing on our side, will probably get to it in a week or so. Thanks for the changes.

👍

marceloverdijk commented:
@srinivasankavitha I've added the wildcard type for lists in interfaces as well. I'm using the generated classes successfully in a local project.

srinivasankavitha commented:
Apologies - haven't gotten around to testing this yet. Will do that later this week.

srinivasankavitha commented on graphql-dgs-codegen-core/src/main/kotlin/com/netflix/graphql/dgs/codegen/generators/java/TypeUtils.kt:
```diff
@@ -51,18 +46,33 @@ class TypeUtils(private val packageName: String, private val config: CodeGenConf
-                val parameterizedTypeName = ParameterizedTypeName.get(ClassName.get(List::class.java), boxed)
+
+                var canUseWildcardType = false
+                if (useWildcardType) {
```
What does useWildcardType indicate?

marceloverdijk replied:
Hi @srinivasankavitha the `useWildcardType` is used to generate `? extends X` like in:

```
public interface IBookPage {
    List<? extends IBook> getItems();
    IPageInfo getPageInfo();
}
```

srinivasankavitha commented:
The changes look good to me, thanks for the PR! Had one minor comment more for my own understanding that I added to the reviews. Also, would you be able to resolve the conflicts? I will merge the PR after that.

marceloverdijk commented:
Hi @srinivasankavitha, thx for the review!
I added a comment for the `useWildcardType` conversation, hope that clarifies it.

This evening I will resolve the conflicts and I will let you know when done.

marceloverdijk commented:
Hi @srinivasankavitha, thx for the review!
I added a comment for the `useWildcardType` conversation, hope that clarifies it.

This evening I will resolve the conflicts and I will let you know when done.

marceloverdijk commented:
Hi @srinivasankavitha I've merged in the updates from `master` and resolved the conflicts.

marceloverdijk commented:
Cool! Looking forward to a new release.
